metadata/buckets_test.go

```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = db.View(func(tx *bolt.Tx) error {
+		if !hasSharedLabel(tx, "testing-with-sharable") {
+			return errors.New("hasSharedLabel should return true when label is set")
+		}
+		if hasSharedLabel(tx, "testing-without-sharable") {
+			return errors.New("hasSharedLabel should return false when label is not set")
+		}
+		return nil
+	})
+
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestGetSharableBucket(t *testing.T) {
+	tmpdir, err := ioutil.TempDir("", "bucket-testing-")
+	if err != nil {
+		t.Error(err)
+	}
+
+	db, err := bolt.Open(filepath.Join(tmpdir, "metadata.db"), 0660, nil)
+	if err != nil {
+		t.Error(err)
+	}
+
+	goodDigest := digest.FromString("gooddigest")
+	imagePresentNS := "has-image-is-sharable"
+	imageAbsentNS := "image-absent"
+
+	// Create two namespaces, empty for now
+	err = db.Update(func(tx *bolt.Tx) error {
+		_, err := createImagesBucket(tx, imagePresentNS)
+		if err != nil {
+			return err
+		}
+
+		_, err = createImagesBucket(tx, imageAbsentNS)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Test that getSharableBucket is correctly returning nothing when a
+	// a bucket with that digest is not present in any namespace.
+	err = db.View(func(tx *bolt.Tx) error {
+		if bkt := getSharableBucket(tx, goodDigest); bkt != nil {
+			return errors.New("getSharableBucket should return nil if digest is not present")
+		}
+		return nil
+	})
+
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Create a blob bucket in one of the namespaces with a well-known digest
+	err = db.Update(func(tx *bolt.Tx) error {
+		_, err = createBlobBucket(tx, imagePresentNS, goodDigest)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Verify that it is still not retrievable if the sharable label is not present
+	err = db.View(func(tx *bolt.Tx) error {
+		if bkt := getSharableBucket(tx, goodDigest); bkt != nil {
+			return errors.New("getSharableBucket should return nil if digest is present but doesn't have sharable label")
+		}
+		return nil
+	})
+
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Create the namespace labels bucket and mark it as sharable
+	err = createNamespaceLabelsBucket(db, imagePresentNS, true)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Verify that this digest is retrievable from getSharableBucket
+	err = db.View(func(tx *bolt.Tx) error {
+		if bkt := getSharableBucket(tx, goodDigest); bkt == nil {
+			return errors.New("getSharableBucket should not return nil if digest is present")
+		}
+		return nil
+	})
+
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func createNamespaceLabelsBucket(db transactor, ns string, sharable bool) error {
+	err := db.Update(func(tx *bolt.Tx) error {
+		err := withNamespacesLabelsBucket(tx, ns, func(bkt *bolt.Bucket) error {
+			if sharable {
+				err := bkt.Put([]byte(sharedNamespaceLabel), []byte("true"))
```

Comment:
This will give the impression that adding the label, and setting it to `"false"` would be the reverse; should we check on `true` / `false` values?

Reply:
I think that makes sense. I will add a check that the value is also set to "true". If it's set to anything else will result in the namespace not being sharable.